File: Lintcode 017 - Subsets.java
Lintcode 017 - Subsets.java

class Solution {
    /**
     * @param S: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    public ArrayList<ArrayList<Integer>> subsets(int[] nums) {
        // write your code here
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        ArrayList<Integer> list = new ArrayList<Integer>();
        
        Arrays.sort(nums);
        dfs(result, list, nums, 0);
        return result;
        
    }
    
    public void dfs (ArrayList<ArrayList<Integer>> result, ArrayList<Integer> list, int[] nums, int pos) {
        
        result.add(new ArrayList<Integer>(list));
        
        for(int i = pos; i < nums.length; i++) {
            list.add(nums[i]);
            dfs(result, list, nums, i + 1);
            list.remove(list.size()-1);
        }
    }
    
    //[[],[1],[1,2],[1,2,3],[1,3],[2],[2,3],[3]]
}

File: Lintcode 017 - Subsets.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace CSharp
{
    class Program{
        static void Main()
        {
            int[] nums = new int[3]{1,3,4};
            List<List<int>> result = new List<List<int>>{};   
            List<int> list = new List<int>{};

            Subset(nums, result, list, 0);

            foreach(var clist in result)
            {
                foreach(var num in clist)
                {
                    Console.Write(num + ", ");
                }
                Console.WriteLine(" ");
            }
        }

        static void Subset(int[] nums, List<List<int>> result, List<int> list, int pos)
        {

            result.Add(new List<int>(list));
            for(int i = pos; i< nums.Length; i++)
            {
                list.Add(nums[i]);
                Subset(nums, result, list, i + 1);
                list.Remove(list[list.Count - 1]);
            }
        }
    }
}
